### ISSUE TITLE
[Feature] InterlockedIncrement/InterlockedIncrement

### ISSUE THREAD

**Issue by Sunbreak** [halildurmus/win32#181]
# Background

[InterlockedIncrement](https://docs.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedincrement) & [InterlockedDecrement](https://docs.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockeddecrement) are required to implement `connectivity_windows` via `dart:ffi`, according to `connectivity_plus_windows`

- Referenced [`connectivity_plus_windows` implementation](https://github.com/fluttercommunity/plus_plugins/blob/79d62dd32579855d8293bf6f07fe4487c1ad5e45/packages/connectivity_plus_windows/windows/network_manager.cpp#L37-L45):

```cpp
  ULONG STDMETHODCALLTYPE AddRef() { return InterlockedIncrement(&lRef); }

  ULONG STDMETHODCALLTYPE Release() {
    LONG lAddend = InterlockedDecrement(&lRef);
    if (lRef == 0) {
      delete this;
    }
    return lAddend;
  }
```

- Refereced [`connectivity_windows`](https://github.com/Sunbreak/plugins/blob/8a36c44f55a46707a4d32a03008d2e12c0a96320/packages/connectivity/connectivity_windows/lib/connectivity_windows.dart#L165-L173) in https://github.com/flutter/plugins/pull/3714

```dart
final _AddRef_Dart onAddRef = (Pointer obj) {
  // TODO InterlockedIncrement(obj.value);
  return 0;
};

final _Release_Dart onReleaseRef = (Pointer obj) {
  // TODO InterlockedDecrement(obj.value);
  return 0;
};
```

# Problem

Add InterlockedIncrement/InterlockedDecrement to win32api.json doen't work

```json
    "InterlockedDecrement": {
        "prototype": "LONG _InterlockedDecrement(\n        LONG *Addend\n      );",
        "dllLibrary": "kernel32",
        "comment": "Decrements (decreases by one) the value of the specified 32-bit variable as an atomic operation."
    },
    "InterlockedIncrement": {
        "prototype": "LONG _InterlockedIncrement(\n        LONG *Addend\n      );",
        "dllLibrary": "kernel32",
        "comment": "Increments (increases by one) the value of the specified 32-bit variable as an atomic operation."
    },
```

According to [InterlockedIncrement](https://docs.microsoft.com/en-us/cpp/intrinsics/interlockedincrement-intrinsic-functions?view=msvc-160) and [InterlockedDecrement](https://docs.microsoft.com/en-us/cpp/intrinsics/interlockeddecrement-intrinsic-functions?view=msvc-160), they are **Microsoft-specific** [Compiler intrinsics](https://docs.microsoft.com/en-us/cpp/intrinsics/compiler-intrinsics?view=msvc-160)

**Comment by timsneath** [comment]
This is beyond my expertise, but I don't _think_ concurrency is an issue here. Dart uses single-threaded isolates, and presumably you're not creating a free-threaded COM server? I don't think there's a risk of reentrancy in an apartment model object. But I'll bet @csells can speak to this with more authority than I have.

**Comment by Sunbreak** [comment]
Do you mean I don't have to implement AddRef/Release as https://github.com/flutter/plugins/pull/3714 shows?

**Comment by csells** [comment]
The underbar versions of these functions are compiler intrinsics, but I believe there are Win32 kernel APIs that do the same thing, e.g. https://docs.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedincrement.

For the STA thread, which you decide by the argument you pass to CoInitializeEx, you can use a simple ++ and -- for AddRef and Release implementations, since COM will queue calls from other apartments via the message queue. If you start passing around objrefs between threads yourself, you'll have trouble but Dart makes that extremely difficult.

tl;dr; try the kernel InterlockedDecrement/InterlockedIncrement if you're worried and fall back on --/++ for STA implementations otherwise.

**Comment by timsneath** [comment]
Unfortunately, the kernel APIs are just a `#define` for the underlying intrinsic. So there's nothing available to project in my package. So like Chris said, I think the best approach is to fall back on the language increment / decrement.

**Comment by Sunbreak** [comment]
I'll try ++/-- for STA, thanks a lot
